src/repeated_game.jl

```diff
@@ -0,0 +1,339 @@
+    r = sqrt(r1 + r2)
+
+    return create_unitcircle_points(rpd, nH, o, r)
+end
+
+#
+# Linear Programming Functions
+#
+"""
+Initialize matrices for the linear programming problems. It sets up the A
+matrix since it never changes, but only allocates space for b and c since
+they will be filled repeatedly on the iterations
+
+We add nH slack variables (which will be constrained to be positive) to
+deal with inequalities associated with Ax \leq b.
+
+min c ⋅ x
+    Ax = b
+
+In this case, the `c` vector will be determined by which subgradient is being
+used, so this function only allocates space for it.
+
+The `A` matrix will be filled with nH set constraints and 2 incentive compatibility
+constraints. The set constraints restrain the linear programming problem to pick
+solutions that are in the current set of continuation values while the incentive
+compatibility constraints ensure the agents won't deviate.
+
+The `b` vector is associated with the `A` matrix and gives the value for constraint.
+"""
+function initialize_LP_matrices(rpd, H)
+    # Need slack variable for every subgradient and additional 2 incentive constraint
+    nH = size(H, 1)
+    nslack = nH + 2
+
+    # Create the c vector (objective)
+    c = zeros(nslack+2)
+
+    # Create the A matrix (constraints)
+    A_H = hcat(H, eye(nH, nslack))
+    A_IC_1 = zeros(1, nslack+2)
+    A_IC_2 = zeros(1, nslack+2)
+    A_IC_1[1] = -rpd.δ
+    A_IC_1[end-1] = 1.0
+    A_IC_2[2] = -rpd.δ
+    A_IC_2[end] = 1.0
+    A = vcat(A_H, A_IC_1, A_IC_2)
+
+    # Create the b vector (constraints)
+    b = Array(Float64, nslack)
+
+    return c, A, b
+end
+
+"""
+Given a constraint w ∈ W, this finds the worst possible payoff for agent i
+
+The output of this function is used to create the values associated with
+incentive compatibility constraints
+"""
+function worst_value_i(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}, i::Int)
+    # Objective depends on which player we are minimizing
+    c = zeros(2)
+    c[i] = 1.0
+
+    # Lower and upper bounds for w
+    lb = [-Inf, -Inf]
+    ub = [Inf, Inf]
+
+    lpout = linprog(c, H, '<', C, lb, ub, ClpSolver())
+    if lpout.status == :Optimal
+        out = lpout.sol[i]
+    else
+        out = minimum(rpd.sg.players[i].payoff_array)
+    end
+
+    return out
+end
+
+worst_value_1(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 1)
+worst_value_2(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 2)
+worst_values(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    (worst_value_1(rpd, H, C), worst_value_2(rpd, H, C))
+
+#
+# Outer Hyper Plane Approximation
+#
+"""
+Approximates the set of equilibrium value set for a repeated game with the
+outer hyperplane approximation described by Judd, Yeltekin, Conklin 2002
+"""
+function outerapproximation(rpd::RepeatedGame; nH=32, tol=1e-8, maxiter=500, nskipprint=1)
```

Comment:
* Currently the function works only for two-player games. So `rpd::RepeatedGame` should be `rpd::RepeatedGame{2}`, or explicitly throw an error if the input game has more than two players.

* In the final version: A `verbose` option should be given, where `verbose=0` would suppress reportings. The default value of `nskipprint` should be much bigger than 1, such as 50 or 100.